### ISSUE TITLE
Using api.get.email_source() fails in MS Edge

### ISSUE THREAD

**Issue by josteink** [KartikTalwar/gmail.js#353]
No really. If you want to write WebExtensions which supports all modern browers, MS Edge is a target too.  :)

Basically something fails because of XHR with bad HTTP methods, and while it may not be fatal, the way we handle it causes it to fail:

* `api.get.email_source()` requests something from https://mail.google.com.
* This results in a 302 redirect.
* The new URL gets probed by jQuery using a HTTP OPTION request.
* This is where it fails. If we could *catch* this error, and just retry with a GET, we should be cool.

You can see this in the developer console here:

![image](https://cloud.githubusercontent.com/assets/411338/21090384/b611bd1a-c03f-11e6-8806-39e3b7f98a95.png)

![image](https://cloud.githubusercontent.com/assets/411338/21090416/e83bba98-c03f-11e6-863d-c682a2c06ac7.png)

@michi88: Anything you think you could contribute on this issue? You've touched this very specific code yourself :)

**Tip:** If you need help enabling extensions in Edge, [check the uBlock Edge docs for how to enable developer mode](https://github.com/nikrolls/uBlock-Edge#installation).

**Comment by josteink** [comment]
This might actually be an issue in Edge itself:

https://developer.microsoft.com/en-us/microsoft-edge/platform/issues/9869121/

**Comment by michi88** [comment]
We did have some CORS issues briefly discussed here: https://github.com/KartikTalwar/gmail.js/pull/333

This is related to the previous redirect being cached by the browser. This request needs to hit google as they will not add CORS headers to the download location otherwise. They open some kind of session or something.

After I've added jQuery's `{cache: false}` which adds something like this to the url `_=76348726487` it started working.

Maybe you can try to disable cache in Edge and see if that make a difference?

**Comment by josteink** [comment]
Status update: From bad to worse.

First: I've now encountered this same behaviour in Firefox too.

Second: According to a network-specialist at my company, this behaviour (issuing a preflight OPTIONS request to check CORS-settings when redirected to a new domain) is actually the *correct* thing to do, and the behaviour in Chrome (which works for us) should be considered a bug.

As long as Gmail keeps it's current behaviour (returning a 405 when getting the OPTIONS request), this means we might soon see this behaviour in all browsers, and thus `email_source()` in all its variants will stop working.

As such, I think it would be wise to investigate a work-around for Firefox and EDGE, because we might actually *need* it to having a working library for Chrome in the not to distant future.

Anyone feel called to investigate this issue further?

cc: @derekribbons, @garris

**Comment by michi88** [comment]
Are you sure `email_source_async` is also not working? If I remember correctly Chrome was issuing a preflight options check for CORS but the CORS headers returned by Google was only present when a cache busting param (from jquery in this case) was added (for some reason). Looking at this now I can confirm that Chrome is NOT doing the CORS OPTIONS request anymore and without the cache busting param it now also works. I'm pretty sure Chrome did do the OPTIONS request as that was the whole problem when debugging this.

**Comment by michi88** [comment]
Shall I wait a bit with deprecating `email_source` https://github.com/KartikTalwar/gmail.js/issues/357 as we might find a fix? I'm still out of ideas though...

**Comment by josteink** [comment]
Yes.

To be absolutely clear on this, I'm exclusively using the `email_source_async` version in my code, and I'm now getting failures with this in Firefox too.

And based on your comment... If you would bother to take a minute and  point me at your concrete findings and data with regard to Chrome's current and past-behaviour (so that I can run that by our local security guy) that would be immensely helpful. :)

**Comment by michi88** [comment]
@josteink Sorry for the late reply. Unfortunately I'm unable to reproduce, or find screenshots which I did make :(, the issue how it was earlier when I was developing the 'fix' for this CORS issue.

Not sure what to do at this point. Do you have any new observations?

What I can image is that chrome does not do the CORS OPTIONS preflight request as they did it already for that domain in some other request. I have no proof for this though.

**Comment by josteink** [comment]
For now, I can get this to work (in *all* browsers) using the following workaround:

1. Use `api.helper.get.email_source_pre()` to get the download URL.
2. Append a random query-parameter with an incrementing counter, thus bust any cache standing in our way (I've established experimentally this to be required for reliable, repeatable operation).
3. Use a "manual" XHR-based download instead of `$.ajax()` (which seems to do way too much for its own good, inserting extra HTTP-headers being one of them). This has the added benefit of letting me download the content as a ArrayBuffer/Uint8Array and bypass any encoding issues downloading MIME as regular text might give.

So for me, this is no longer an issue.

It does leave me with the question though, if we should update the general `gmail.js` implementation for `api.get.email_source()` with this approach?

While my workarounds works for me in my projects, I'm not entirely sure if they are a 100% appropriate drop-in replacement for whatever people has built on top on this function already.

Opinions?

cc: @KartikTalwar

**Comment by michi88** [comment]
As we've added `email_source_async` pretty recently and we can easily test if it works or not I think we should implement your manual XHR-based download method. For me it works now but I only target Chrome at the moment.

> I can provide details on why this seems to be needed, if anyone is interested.

I only know that it works and was baffled by it... so please enlighten me :)

BTW, I think the current `email_source` implementation is beyond repair unfortunately. As discussed here https://github.com/KartikTalwar/gmail.js/issues/357 we should probably deprecate it.

**Comment by KartikTalwar** [comment]
Sorry for jumping in late here but I what @josteink mentioned in his workaround makes sense to me. After inspecting this, here is a bit more detail on what's going on (for future readers):

1. Preflight check will happen for all requests originating from client side
2. `api.helper.get.email_source_pre()` generates a url for the email data
3. the url from 2. is actually a pointer (redirects) to the real email data url
   - This real data is hosted on the domain `mail-attachment.googleusercontent.com`
   - The `email_source_pre` url belongs to the domain `mail.google.com`
4. The CORS headers do not allow `mail.google.com` to access `mail-attach...` and reject the request
5. jQuery's `ajax()` does the preflight and on success runs the actual `GET`
   - Since the preflight fails for us, the GET does too
6. Re: cache-busting, I believe its chrome being smart by inspecting the cache headers on the url and determining client-side that the server shouldn't be pinged again. Adding a random param to the url will make chrome treat it as a new url

The real solution in my opinion is to find another way of getting the email data - either by always inspecting the requests or finding another url that belongs on mail.google.com but using native XHR for now is a good solution. 

WRT compatibility, if we can make that function return the data synchronously (without having the library deal incorporating all the callback structure/overhead), that'd be a win, otherwise we should retire the function. 

Attached:

<img width="918" alt="screen shot 2017-02-05 at 4 36 24 pm" src="https://cloud.githubusercontent.com/assets/461702/22631402/7cc42950-ebc1-11e6-9af9-18b8e362e638.png">

**Comment by josteink** [comment]
> WRT compatibility, if we can make that function return the data synchronously (without having the library deal with/incorporating all the callback structure/overhead), that'd be a win, otherwise we should retire the function.

That seems a bit rash IMO. Allow me argue why.

## Overview

Oh noes. I've added headers. It seems this will be a long message :)

When we want to download the raw MIME-message, the URL we generate (and which we are allowed to make, as a script running on `https://mail.google.com/`, is to another resource on the same domain.

As @KartikTalwar points out, that's not where the data is actually stored, so we are redirected by a HTTP 302 to the actual domain hosting the data: `mail-attachment.googleusercontent.com`.

This means we are crossing a domain-boundary, but since the request *originated* on `mail.google.com`, that's OK, as long as the receiving host (`mail-attachment.googleusercontent.com`) is also willing to accept our request.

This is what the preflight check is doing. It's checking for CORS headers, which will list `mail.google.com` as a valid source for requests. If this header is present in the response, the browser will proceed with the actual, real request.

So for all good, all standard. All browsers are (or should be) doing this. It's pretty much the spec for how things should work. (Or at least so I've been told).

## Issues

Now this is where things get tricky, because we have 3 different issues, 2 of them which can easily get conflated, and those are both related to caching:

* 405 response from `mail-attachment.googleusercontent.com` because of bad request.
* 401 response from `mail-attachment.googleusercontent.com` because of invalid authentication cookies.
* Character encoding issues due to downloading as text.

The third issue is not really *directly* related to the others, but if we're making any changes or API-changes in response to the first ones, it should be considered too, so I'm including it anyway :)

## Issue 1: Bad request 

Let's tackle the bad request first:

The workaround we have for this right now, is sending `{cache: false}` to `jQuery.ajax()`, but that unfortunately only seems to work in Chrome right now.

> I only know that it works and was baffled by it... so please enlighten me :)

From what I can tell, `jQuery.ajax()` is attempting to speed things up by exploiting cached content. Towards this goal, it might add additional HTTP-headers:

https://github.com/jquery/jquery/blob/1.10.2/jquery.js#L8048-L8080

My guess is that `mail-attachment.googleusercontent.com` isn't expecting these headers and thus responds with "405 Bad Request".

By manually implementing a XHR downloader which does none of that fancy stuff jQuery does, I've been able to completely bypass this problem. As far as I can tell, it only exists because of jQuery. So let's not use jQuery :)

## Issue 2: Failing "unauthenticated" requests.

And when we solve that, it eventually leads us to the 401 unathenticated response...

So before diving into the specifics, just to get it over with, let's remind ourselves of some basics: Just like I can't take (or guess!) a URL from your Gmail, run it in my browser and grab your email, Google does the same for it's raw MIME-messages and email-attachments.

To get any of your personal data, I need to be authenticated to Gmail as you. On the web, this is done through the use of cookies.

Yes I know you *knew* that. But it's easy to forget. And now we all have it in mind, and will give it proper attention. So let's proceed with the actual issue.

When trying to (manually via XHR) download the `email_source`, the initial response will succeed for a single given `emailID`. And it will succeed for other emailIDs too.

But if you try to request the `email_source` for the same `emailID` more than once, things will start failing. With a 401.

To find the reason for this, we need to inspect the actual browser requests and responses:

The request get's redirected, and as we can see at this point, we have lots of cookies!

![image](https://cloud.githubusercontent.com/assets/411338/22637352/b3f688d4-ec41-11e6-81c3-7dd8e85585c1.png)

One of those cookies (or several, in combination) represents our authenticated login to Gmail.

We follow the redirect to `mail-attachment.googleusercontent.com`, and get a new response. This response however tells us to go obsolete some of our cookies. Immediately!

**NOTE: The following request was issued at 7:53:03 GMT, despite the misleading `Date:` and `Expired:` headers:**

![image](https://cloud.githubusercontent.com/assets/411338/22637477/3f9ceae0-ec42-11e6-8895-4a6d7aa0fda3.png)

That means that if we repeat the original request (unmodified) to `mail.google.com`, the browser will most likely remember the 302, and go directly to `mail-attachment.googleusercontent.com`.

But if we re-request the same resource on `mail-attachment.googleusercontent.com`, we will now have a *different* set of cookies, since we obsoleted some on our last request!

And so, it seems we're suddenly unauthenticated!

## A few observations:

I've observed the following:

* Using manual XHR over jQuery.ajax() does less "sophisticated" tricks, and seems to work consistently better. 
* Our cookies getting obsoleted *immediately* may partially explain why we have observed different behaviour on sync-vs-async requests?
* Applying a cache-busting query-parameter seems to somehow bypass the issue of getting unauthenticated requests once we actually manage to cross the CORS barrier, although I haven't been able to determine why. That is: I don't see where we get our cookies re-set to a valid value.
* Also: Downloading MIME-messages as string/text-data may introduce text-encoding issues when passed along to a real MIME-parser, as nested messages may have had their original encoding overridden (For instance: MIME message with message originaly in Windows-1252, downloaded, decoded, passed to a string in unicode form. Passed along as a base-64 encoded UTF-8 stream to a real MIME-engine... Which then tries to decode non-latin letters in a Windows 1252 section, but which are now UTF-8. Ooops)..

In conclusion:

* Let's remove jQuery in favour of manual XHR downloads.
* We need to implement manual cache-busting to allow repeatable downloads.
* To avoid character-encoding issues with email-source, we should have an API/option to permit binary downloads (Uint8Array).

All in agreement say "yes!", and I'll do my best to make it so :)

**Comment by michi88** [comment]
yes!

> Applying a cache-busting query-parameter seems to somehow bypass the issue of getting unauthenticated requests once we actually manage to cross the CORS barrier, although I haven't been able to determine why. That is: I don't see where we get our cookies re-set to a valid value.

Isn't that because the browser will not use the cached redirect response from mail.google.com to mail-attachment.googleusercontent.com when the url is different (cache busting param)?

Also note that if you have devtools open and have `Disable cache` enabled you get different behaviour. Important when we are testing this.


> From what I can tell, jQuery.ajax() is attempting to speed things up by exploiting cached content. Towards this goal, it might add additional HTTP-headers:

After reviewing the jquery source a bit it seems they set the `_=` query param just to bust their own cache... Also, I think it's strange they set the `If-Modified-Since` etc. in all cases (I think they should not do that when `s.cache === false`). This might actually be the source of the problem. Good find @josteink!

**Comment by KartikTalwar** [comment]
@josteink I think we're on the same page here - I meant we should retire the sync version in favor of async as I assume using xhr route will return the email data asynchronously :-)

**Comment by josteink** [comment]
@michi88 welcome to the club!

@KartikTalwar i was thinking about suggesting him as a contributor for a long time, but it seems I forgot. Good choice IMO :)

**Comment by KartikTalwar** [comment]
I'm really glad to have you all amazing contributors 






On Mon, Feb 6, 2017 at 10:09 PM -0800, "Jostein Kjønigsen" <notifications@github.com> wrote:












@michi88 welcome to the club!


@KartikTalwar i was thinking about suggesting him as a contributor for a long time, but it seems I forgot. Good choice IMO :)

**Comment by michi88** [comment]
@KartikTalwar @josteink Thanks guys!

I was really happy when I found this project so I love to contribute where I can.

I have to be honest though, my time is limited. I'll do my best!

**Comment by josteink** [comment]
I've promised to get this done, and I will. But the "production quality" code I have is at the office, so it'll have to wait until I have a "quiet" week there and remember to get this done.

Maybe next week sometime :)

**Comment by josteink** [comment]
Just as (yet another) comment on this issue. I haven't forgotten about it. I've just wanted to harden and double-check that the code I have worked 100% properly as it should.

I've had one very odd intermittent bug which has been bothering me, but this week (I think) I was finally able to sort it out.

That leaves what remains (in my opinion)  good and simple production-quality code. Looking at my schedule, I'll probably have time to get it submitted next week.

And then we can close this issue and proudly say we support all modern browsers (sorry Safari!) :)

**Comment by josteink** [comment]
And there we go. Fix is live and ready for review :)
